<a href="https://colab.research.google.com/github/danielburch/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module3-join-and-reshape-data/LS_DS_113_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 3*

---

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [37]:
#housekeeping

%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns


pd.set_option('display.max_rows', 150)
pd.set_option('display.max_columns', 100)

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [1]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2020-06-03 19:43:11--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.113.125
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.113.125|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  55.2MB/s    in 3.7s    

2020-06-03 19:43:15 (53.0 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [3]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [4]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [5]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [0]:
##### YOUR CODE HERE #####

'''
Columns needed:
 - order_id (found in order_products)
 - product_name (found in products.csv)

 I need to count the order_ids that contain the product_names

'''

In [6]:
# Explore the orders.csv
orders = pd.read_csv('orders.csv')
print (orders.shape)
orders.head()

(3421083, 7)


order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [7]:
# Explore the products.csv
products = pd.read_csv('products.csv')
print(products.shape)
products.head()

(49688, 4)


product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  
0             19  
1             13  
2              7  
3              1  
4             13

In [9]:
# Explore the order_products__prior

order_products__prior = pd.read_csv('order_products__prior.csv')
print(order_products__prior.shape)
order_products__prior.head()

(32434489, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [10]:
# Explore the order_products__train

order_products__train = pd.read_csv('order_products__train.csv')
print(order_products__train.shape)
order_products__train.head()

(1384617, 4)


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [0]:
'''
First of all, I don't need the orders df. I don't need any columns from it. I can get all the info 
I need from the order_products table

I'll join the the products and orders table by using the order_products dataforms. 
First, I will need to combine both order_products dataforms in order to have all data in one dataform
'''

In [13]:
#Joining the order_products__prior and order_products__train 

order_products = pd.concat([order_products__prior, order_products__train])
print(order_products.shape)
order_products.head()

(33819106, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [23]:
#Next I will filter the products df to only include the product_names from the top ten.

top_ten = ['Banana','Bag of Organic Bananas','Organic Strawberries','Organic Baby Spinach',
                                          'Organic Hass Avocado','Organic Avocado','Large Lemon','Strawberries','Limes',
                                          'Organic Whole Milk']

products_subset = products[products['product_name'].isin(top_ten)]

products_subset.head(20)

product_id            product_name  aisle_id  department_id
13175       13176  Bag of Organic Bananas        24              4
16796       16797            Strawberries        24              4
21136       21137    Organic Strawberries        24              4
21902       21903    Organic Baby Spinach       123              4
24851       24852                  Banana        24              4
26208       26209                   Limes        24              4
27844       27845      Organic Whole Milk        84             16
47208       47209    Organic Hass Avocado        24              4
47625       47626             Large Lemon        24              4
47765       47766         Organic Avocado        24              4

In [31]:
#Next, I'll clean up the columns in products_subset
products_subset = products_subset.drop(['aisle_id','department_id'], axis =1)

products_subset.head(5)

product_id            product_name
13175       13176  Bag of Organic Bananas
16796       16797            Strawberries
21136       21137    Organic Strawberries
21902       21903    Organic Baby Spinach
24851       24852                  Banana

In [32]:
#Next I will filter the order_products df by the 10 product_ids found in products_subset
order_products_subset = order_products[order_products['product_id'].isin(products_subset['product_id'])]

order_products_subset.head()

order_id  product_id  add_to_cart_order  reordered
12         3       21903                  4          1
30         5       13176                  1          1
51         5       47209                 22          0
77        10       24852                  1          1
80        10       47766                  4          1

In [33]:
#Next I'll clean up the columns of order_products_subset

order_products_subset = order_products_subset.drop(['add_to_cart_order','reordered'], axis =1)

order_products_subset.head(10)

order_id  product_id
12          3       21903
30          5       13176
51          5       47209
77         10       24852
80         10       47766
83         10       21137
126        14       27845
170        18       47766
180        20       24852
190        22       24852

In [34]:
# Time to merge the products_subset with the order_products_subset

order_products_and_products = pd.merge(products_subset,order_products_subset, on = 'product_id', how = 'inner')

order_products_and_products.head()

product_id            product_name  order_id
0       13176  Bag of Organic Bananas         5
1       13176  Bag of Organic Bananas        27
2       13176  Bag of Organic Bananas        29
3       13176  Bag of Organic Bananas        32
4       13176  Bag of Organic Bananas        42

In [35]:
# Checking on my last 5 rows
order_products_and_products.tail()

product_id     product_name  order_id
2418309       47766  Organic Avocado   3419827
2418310       47766  Organic Avocado   3420232
2418311       47766  Organic Avocado   3420257
2418312       47766  Organic Avocado   3420578
2418313       47766  Organic Avocado   3420857

In [36]:
# Time to count the product_name based on the number of rows created from unique order_id
order_products_and_products['product_name'].value_counts(dropna=False).head(15)

Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
Name: product_name, dtype: int64

**These are the top 10 most frequently ordered products.** 

1. Banana                    **491291**
2. Bag of Organic Bananas    **394930**
3. Organic Strawberries      **275577**
4. Organic Baby Spinach      **251705**
5. Organic Hass Avocado      **220877**
6. Organic Avocado           **184224**
7. Large Lemon               **160792**
8. Strawberries              **149445**
9. Limes                     **146660**
10. Organic Whole Milk        **142813**














## Reshape Data Section

- Replicate the lesson code. Remember, if you haven't followed along typing out what we did during lecture, do that now to get more repetition with the syntax.
- Take table 2 (the transpose of table 1) and practice changing it into Tidy format and back again. You should not use the transpose operation anywhere in this code:
 - Table 2 --> Tidy
 - Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [90]:
import pandas as pd
import numpy as np

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [91]:
# table2 -> tidy

# First step is to use the pivot_table to adjust the indices and columns

#table2 = table2.set_index(['John Smith','Jane Doe', 'Mary Johnson'], append=False)

table2 = table2.reset_index()

table2




index  John Smith  Jane Doe  Mary Johnson
0  treatmenta         NaN      16.0           3.0
1  treatmentb         2.0      11.0           1.0

In [92]:
tidy = table2.melt(id_vars='index', value_vars=['John Smith', 'Jane Doe', 'Mary Johnson'])

tidy

index      variable  value
0  treatmenta    John Smith    NaN
1  treatmentb    John Smith    2.0
2  treatmenta      Jane Doe   16.0
3  treatmentb      Jane Doe   11.0
4  treatmenta  Mary Johnson    3.0
5  treatmentb  Mary Johnson    1.0

In [93]:
#Reorder columns
reorder = tidy[['variable','index','value']]

reorder

variable       index  value
0    John Smith  treatmenta    NaN
1    John Smith  treatmentb    2.0
2      Jane Doe  treatmenta   16.0
3      Jane Doe  treatmentb   11.0
4  Mary Johnson  treatmenta    3.0
5  Mary Johnson  treatmentb    1.0

In [96]:
#Rename columns
tidy1 = reorder.rename(columns={'variable': 'name', 'index': 'trt','value':'result'})

rename



name         trt  result
0    John Smith  treatmenta     NaN
1    John Smith  treatmentb     2.0
2      Jane Doe  treatmenta    16.0
3      Jane Doe  treatmentb    11.0
4  Mary Johnson  treatmenta     3.0
5  Mary Johnson  treatmentb     1.0

In [97]:
#Get rid of the word treatment

tidy1.trt = rename.trt.str.replace('treatment', '')

tidy1

name trt  result
0    John Smith   a     NaN
1    John Smith   b     2.0
2      Jane Doe   a    16.0
3      Jane Doe   b    11.0
4  Mary Johnson   a     3.0
5  Mary Johnson   b     1.0

In [98]:
# tidy -> table2
wide = tidy1.pivot_table(index='name', columns='trt', values='result')

wide

trt              a     b
name                    
Jane Doe      16.0  11.0
John Smith     NaN   2.0
Mary Johnson   3.0   1.0

In [0]:
flights = sns.load_dataset('flights')

In [0]:
# Flights Pivot Table
'''
Load seaborn's flights dataset by running the cell below. 
Then create a pivot table showing the number of passengers by month and year. 
Use year for the index and month for the columns. 
You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.
'''



In [109]:
flights.head()

year     month  passengers
0  1949   January         112
1  1949  February         118
2  1949     March         132
3  1949     April         129
4  1949       May         121

In [112]:
# Moving year column to index


flights = flights.pivot_table(index='year',columns='month',values='passengers')

flights.head()

month  January  February  March  April  May  June  July  August  September  \
year                                                                         
1949       112       118    132    129  121   135   148     148        136   
1950       115       126    141    135  125   149   170     170        158   
1951       145       150    178    163  172   178   199     199        184   
1952       171       180    193    181  183   218   230     242        209   
1953       196       196    236    235  229   243   264     272        237   

month  October  November  December  
year                                
1949       119       104       118  
1950       133       114       140  
1951       162       146       166  
1952       191       172       194  
1953       211       180       201

In [113]:
flights.tail()

month  January  February  March  April  May  June  July  August  September  \
year                                                                         
1956       284       277    317    313  318   374   413     405        355   
1957       315       301    356    348  355   422   465     467        404   
1958       340       318    362    348  363   435   491     505        404   
1959       360       342    406    396  420   472   548     559        463   
1960       417       391    419    461  472   535   622     606        508   

month  October  November  December  
year                                
1956       306       271       306  
1957       347       305       336  
1958       359       310       337  
1959       407       362       405  
1960       461       390       432

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [115]:
##### YOUR CODE HERE #####

# Visually see what I am buiding

from IPython.display import display, Image
url = 'https://cdn-images-1.medium.com/max/1600/1*wKfV6OV-_1Ipwrl7AjjSuw.png'
example = Image(url=url, width=600)

display(example)